subsys/bluetooth/host/classic/hfp_ag.c

```diff
@@ -385,256 +357,1118 @@ static int hfp_ag_send_data(struct bt_hfp_ag *ag, bt_hfp_ag_tx_cb_t cb, void *us
+
+	return err;
 }
 
-static int get_number(struct net_buf *buf, uint32_t *number)
+static int hfp_ag_force_update_indicator(struct bt_hfp_ag *ag, enum bt_hfp_ag_indicator index,
+					 uint8_t value, bt_hfp_ag_tx_cb_t cb, void *user_data)
 {
-	int err = -EINVAL;
+	int err;
+	uint8_t old_value;
 
-	*number = 0;
+	hfp_ag_lock(ag);
+	old_value = ag->indicator_value[index];
+	ag->indicator_value[index] = value;
+	hfp_ag_unlock(ag);
 
-	skip_space(buf);
-	while (buf->len > 0) {
-		if ((buf->data[0] >= '0') && (buf->data[0] <= '9')) {
-			*number = *number * 10 + buf->data[0] - '0';
-			(void)net_buf_pull(buf, 1);
-			err = 0;
-		} else {
-			break;
-		}
+	LOG_DBG("indicator %d, old %d -> new %d", index, old_value, value);
+
+	err = hfp_ag_send_data(ag, cb, user_data, "\r\n+CIEV:%d,%d\r\n", (uint8_t)index + 1, value);
+	if (err) {
+		hfp_ag_lock(ag);
+		ag->indicator_value[index] = old_value;
+		hfp_ag_unlock(ag);
+		LOG_ERR("Fail to update indicator %d, current %d", index, old_value);
 	}
-	skip_space(buf);
 
 	return err;
 }
 
-static bool is_char(struct net_buf *buf, uint8_t c)
+static void clear_call_setup_ind_cb(struct bt_hfp_ag *ag, void *user_data)
 {
-	bool found = false;
+	uint8_t call_ind;
+	int err;
 
-	skip_space(buf);
-	if (buf->len > 0) {
-		if (buf->data[0] == c) {
-			(void)net_buf_pull(buf, 1);
-			found = true;
+	hfp_ag_lock(ag);
+	call_ind = ag->indicator_value[BT_HFP_AG_CALL_IND];
+	hfp_ag_unlock(ag);
+
+	if (call_ind) {
+		err = hfp_ag_update_indicator(ag, BT_HFP_AG_CALL_IND, 0, NULL, NULL);
+		if (err) {
+			LOG_ERR("Fail to clear call_setup ind");
 		}
 	}
-	skip_space(buf);
+}
 
-	return found;
+static void clear_call_ind_cb(struct bt_hfp_ag *ag, void *user_data)
+{
+	uint8_t call_setup_ind;
+	int err;
+
+	hfp_ag_lock(ag);
+	call_setup_ind = ag->indicator_value[BT_HFP_AG_CALL_SETUP_IND];
+	hfp_ag_unlock(ag);
+
+	if (call_setup_ind) {
+		err = hfp_ag_update_indicator(ag, BT_HFP_AG_CALL_SETUP_IND, BT_HFP_CALL_SETUP_NONE,
+					      NULL, NULL);
+		if (err) {
+			LOG_ERR("Fail to clear call_setup ind");
+		}
+	}
 }
 
-static int bt_hfp_ag_brsf_handler(struct bt_hfp_ag *ag, struct net_buf *buf)
+static void clear_call_and_call_setup_ind(struct bt_hfp_ag *ag)
 {
-	uint32_t hf_features;
+	uint8_t call_setup_ind;
+	uint8_t call_ind;
 	int err;
 
-	if (!is_char(buf, '=')) {
-		return -ENOTSUP;
+	hfp_ag_lock(ag);
+	call_setup_ind = ag->indicator_value[BT_HFP_AG_CALL_SETUP_IND];
+	call_ind = ag->indicator_value[BT_HFP_AG_CALL_IND];
+	hfp_ag_unlock(ag);
+
+	if (call_setup_ind) {
+		err = hfp_ag_update_indicator(ag, BT_HFP_AG_CALL_SETUP_IND, BT_HFP_CALL_SETUP_NONE,
+					      clear_call_setup_ind_cb, NULL);
+		if (err) {
+			LOG_ERR("Fail to clear call_setup ind");
+		} else {
+			return;
+		}
 	}
 
-	err = get_number(buf, &hf_features);
-	if (err != 0) {
-		return -ENOTSUP;
+	if (call_ind) {
+		err = hfp_ag_update_indicator(ag, BT_HFP_AG_CALL_IND, 0, clear_call_ind_cb, NULL);
+		if (err) {
+			LOG_ERR("Fail to clear call ind");
+		}
+	}
```

Comment:
Could you add some code comments here to explain the logic of this? It's not clear why you're bailing out from the function if `hfp_ag_update_indicator(ag, BT_HFP_AG_CALL_SETUP_IND, ...)` succeeds, but proceeding to call  `hfp_ag_update_indicator(ag, BT_HFP_AG_CALL_IND, ...)` if the first call fails.

Reply:
These two values ​​are usually not 0 at the same time. But there is an corner case that, the CALL=1, CALL_SETUP is not zero. This is usually the case when something goes wrong. So, I did not notice it is an issue. I will update it.

Reply:
Removed the line 483